File: 프로그래머스/동적계획법/등굣길.java
//동적계획법_3번_등굣길

package 프로그래머스.동적계획법;

import java.util.*;

public class 등굣길 {
    public static void main(String args[]){
        int m = 4;
        int n = 3;
        int[][] puddles = {{2,2}};

        System.out.println(solution(m, n, puddles));
    }

    public static int div = 1000000007;
    public static int max = Integer.MAX_VALUE;
    public static long[][] dp;
    public static long[][] dist;
    public static int solution(int m, int n, int[][] puddles) {
        int answer = 0;
        dp = new long[n+1][m+1];
        for(int[] pud : puddles){
            dp[pud[1]][pud[0]] = max;
        }
        dist = new long[n+1][m+1];
        for(int i=0;i<=n;i++){
            Arrays.fill(dist[i], max);
        }

        dp[1][1] = 1;
        dist[1][1] = 0;
        for(int i=1;i<=n;i++){
            for(int j=1;j<=m;j++){
                if(dp[i][j]!=0) continue;
                
                dist[i][j] = Math.min(dist[i][j-1], dist[i-1][j])+1;
                if(dist[i][j]==dist[i][j-1]+1 && dp[i][j-1]!=max){
                    dp[i][j]+=dp[i][j-1];
                }
                if(dist[i][j]==dist[i-1][j]+1 && dp[i-1][j]!=max){
                    dp[i][j]+=dp[i-1][j];
                }
                if(dp[i][j]>div) dp[i][j]%=div;
            }
        }

        answer = (int)(dp[n][m]%div);
        return answer;
    }
    
}

File: 프로그래머스/동적계획법/등굣길.py
dx = [1, 0]
dy = [0, 1]
def solution(m, n, puddles):
    dp = [[-1 for _ in range(m)] for _ in range(n)]
    return (DFS(0, 0, n, m, puddles, dp))
    

def DFS(x, y, n, m, puddles, dp):
    
    if x==n-1 and y==m-1:
        return 1
    
    if dp[x][y]!=-1:
        return dp[x][y]
    
    dp[x][y] = 0
    for i in range(2):
        nx = x+dx[i]
        ny = y+dy[i]
        if nx<0 or ny<0 or nx>=n or ny>=m:
            continue
        if [nx, ny] in puddles:
            continue
        dp[x][y] += DFS(nx, ny, n, m, puddles, dp)
    
    return dp[x][y] % 1000000007

print(solution(4, 3, [[2, 2]]))